File: src/api/controllers/invoices.controller.ts
import { Request, Response, NextFunction } from "express";
import { z } from "zod";
import { InvoiceStatus } from "@prisma/client";
import {
  createInvoice,
  getInvoice,
  listInvoices,
  cancelInvoice,
  CreateInvoiceSchema,
} from "../../services/invoice.service";
import {
  enqueueSign,
  enqueueSend,
  enqueueCheckAuth,
} from "../queues/billing.queue.proxy";
import { logger } from "../../utils/logger";

// ─── GET /invoices ─────────────────────────────────────────────────
export async function listInvoicesHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const status = req.query.status as InvoiceStatus | undefined;
    const page = Number(req.query.page ?? 1);
    const limit = Number(req.query.limit ?? 20);

    const result = await listInvoices({ status, page, limit });
    res.json({ success: true, ...result });
  } catch (err) {
    next(err);
  }
}

// ─── POST /invoices ────────────────────────────────────────────────
export async function createInvoiceHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const input = CreateInvoiceSchema.parse(req.body);
    const invoice = await createInvoice(input);

    // Encolar firma automáticamente si hay certificado asignado
    if (invoice.certificateId) {
      await enqueueSign(invoice.id);
      logger.info("Job de firma encolado", { invoiceId: invoice.id });
    }

    res.status(201).json({
      success: true,
      invoice,
      message: invoice.certificateId
        ? "Factura creada. Firma y envío encolados automáticamente."
        : "Factura creada. Asigne un certificado y llame a POST /invoices/:id/sign.",
    });
  } catch (err) {
    next(err);
  }
}

// ─── GET /invoices/:id ────────────────────────────────────────────
export async function getInvoiceHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const invoice = await getInvoice(req.params.id);
    res.json({ success: true, invoice });
  } catch (err) {
    next(err);
  }
}

// ─── POST /invoices/:id/sign ──────────────────────────────────────
export async function signInvoiceHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const job = await enqueueSign(req.params.id);
    res.json({
      success: true,
      message: "Job de firma encolado",
      jobId: job.id,
    });
  } catch (err) {
    next(err);
  }
}

// ─── POST /invoices/:id/send ──────────────────────────────────────
export async function sendInvoiceHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const job = await enqueueSend(req.params.id);
    res.json({
      success: true,
      message: "Job de envío al SRI encolado",
      jobId: job.id,
    });
  } catch (err) {
    next(err);
  }
}

// ─── POST /invoices/:id/check-auth ───────────────────────────────
export async function checkAuthHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const job = await enqueueCheckAuth(req.params.id, 0);
    res.json({
      success: true,
      message: "Consulta de autorización encolada",
      jobId: job.id,
    });
  } catch (err) {
    next(err);
  }
}

// ─── DELETE /invoices/:id ─────────────────────────────────────────
export async function cancelInvoiceHandler(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const { reason } = z.object({ reason: z.string().optional() }).parse(req.body);
    const invoice = await cancelInvoice(req.params.id, reason);
    res.json({ success: true, invoice });
  } catch (err) {
    next(err);
  }
}

File: src/VenturacomSri.Api/Controllers/InvoicesController.cs
using Hangfire;
using Microsoft.AspNetCore.Mvc;
using VenturacomSri.Api.DTOs;
using VenturacomSri.Api.Jobs;
using VenturacomSri.Api.Models;
using VenturacomSri.Api.Services;

namespace VenturacomSri.Api.Controllers;

[ApiController]
[Route("api/v1/[controller]")]
public class InvoicesController : ControllerBase
{
    private readonly InvoiceService _svc;

    public InvoicesController(InvoiceService svc) => _svc = svc;

    // GET /api/v1/invoices?status=Draft&page=1&limit=20
    [HttpGet]
    public async Task<IActionResult> List(
        [FromQuery] InvoiceStatus? status,
        [FromQuery] int page  = 1,
        [FromQuery] int limit = 20)
    {
        limit = Math.Clamp(limit, 1, 100);
        var (total, items) = await _svc.ListAsync(status, page, limit);
        return Ok(new { success = true, total, page, limit, invoices = items });
    }

    // POST /api/v1/invoices
    [HttpPost]
    public async Task<IActionResult> Create([FromBody] CreateInvoiceRequest req)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        var invoice = await _svc.CreateAsync(req);

        string? jobId = null;
        string message;

        if (!string.IsNullOrEmpty(invoice.CertificateId))
        {
            jobId = BackgroundJob.Enqueue<SignInvoiceJob>(j => j.ExecuteAsync(invoice.Id));
            message = "Factura creada. Firma y envío encolados automáticamente.";
        }
        else
        {
            message = "Factura creada. Asigne un certificado y llame a POST /invoices/{id}/sign.";
        }

        return CreatedAtAction(nameof(Get), new { id = invoice.Id },
            new { success = true, invoice, jobId, message });
    }

    // GET /api/v1/invoices/{id}
    [HttpGet("{id}")]
    public async Task<IActionResult> Get(string id)
    {
        var invoice = await _svc.GetAsync(id);
        return Ok(new { success = true, invoice });
    }

    // POST /api/v1/invoices/{id}/sign
    [HttpPost("{id}/sign")]
    public IActionResult Sign(string id)
    {
        var jobId = BackgroundJob.Enqueue<SignInvoiceJob>(j => j.ExecuteAsync(id));
        return Ok(new { success = true, message = "Job de firma encolado.", jobId });
    }

    // POST /api/v1/invoices/{id}/send
    [HttpPost("{id}/send")]
    public IActionResult Send(string id)
    {
        var jobId = BackgroundJob.Enqueue<SendInvoiceJob>(j => j.ExecuteAsync(id));
        return Ok(new { success = true, message = "Job de envío al SRI encolado.", jobId });
    }

    // POST /api/v1/invoices/{id}/check-auth
    [HttpPost("{id}/check-auth")]
    public IActionResult CheckAuth(string id)
    {
        var jobId = BackgroundJob.Enqueue<CheckAuthorizationJob>(j => j.ExecuteAsync(id, 1));
        return Ok(new { success = true, message = "Consulta de autorización encolada.", jobId });
    }

    // DELETE /api/v1/invoices/{id}
    [HttpDelete("{id}")]
    public async Task<IActionResult> Cancel(string id, [FromBody] CancelInvoiceRequest? req)
    {
        var invoice = await _svc.CancelAsync(id, req?.Reason);
        return Ok(new { success = true, invoice });
    }
}
